### PR TITLE
AAP-65395: Add dispatcherd health check to ping endpoint

### PR DESCRIPTION
**Pull request by tznamena**
## Description

- What is being changed?
  - Ping endpoint is updated to show status of dispatcherd
- Why is this change needed?
  - To provide visibility of dispatcherd status
- How does this change address the issue?
  - Add dispatcherd_connected boolean field to the ping response. The check uses dispatcherd's alive control command and is informational only — it does not affect overall status.

Assisted-by: Claude Code claude-opus-4-6 1M context

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Steps to Test
1. Make sure you have latest code and then pull down the changes in this PR
2. Start the gateway
3. Check the ping endpoint output

### Expected Results

It should show:
```json
{
    "status": "good",
    "version": "0.1",
    "pong": "2026-05-08 14:56:30.875622",
    "db_connected": true,
    "proxy_connected": true,
    "dispatcherd_connected": true
}
```

## Summary by CodeRabbit

* **New Features**
  * Status/ping API now reports a dispatcherd connectivity flag; connectivity failures are logged and treated as informational (they do not affect overall health status).

* **Tests**
  * Added and updated ping tests to cover dispatcherd-connected and dispatcherd-down scenarios and to integrate the new connectivity check into existing health checks.

### CODE REVIEW

File: aap_gateway_api/views/api/v1/ping.py [hunk 1]
```diff
@@ -60,5 +65,12 @@ def get(self, request):
                 {'status': STATUS_DEGRADED, 'db_connected': False, 'db_exception': type(e).__name__, 'proxy_exception_type': 'Skipped, DB unavailable.'}
             )
 
+        # Dispatcherd check is informational only — does not affect overall status
+        try:
+            self._check_dispatcherd()
+            response['dispatcherd_connected'] = True
+        except Exception:
+            response['dispatcherd_connected'] = False
```
Comment (Dostonbek1) [inline aap_gateway_api/views/api/v1/ping.py:78]: The `except Exception` block silently swallows failures with no logging. For an informational health check, operators still need to know when dispatcherd is unreachable for debugging. The proxy check already logs `type(e).__name__` — this should follow the same pattern.

```suggestion
        except Exception:
            logger.warning("Dispatcherd health check failed", exc_info=True)
            response['dispatcherd_connected'] = False
```

You'll also need to add `import logging` and `logger = logging.getLogger(__name__)` at the top of the file.

File: aap_gateway_api/views/api/v1/ping.py [hunk 2]
```diff
@@ -20,6 +20,11 @@ def _check_db(self):
         with connection.cursor() as cursor:
             cursor.execute("SELECT 1")
 
+    def _check_dispatcherd(self):
+        from dispatcherd.factories import get_control_from_settings
+
+        get_control_from_settings().control_with_reply("alive", timeout=getattr(settings, "DISPATCHERD_HEALTH_CHECK_TIMEOUT", 5))
```
Comment (Dostonbek1) [inline aap_gateway_api/views/api/v1/ping.py:30]: The return value of `control_with_reply("alive")` is discarded. The caller in `get()` determines connectivity solely by whether an exception is raised. If the dispatcherd `alive` command can ever return an explicit unhealthy/falsy response without throwing, this would produce a false positive.

If the API contract guarantees exceptions on failure, this is fine — but it's worth verifying and adding a brief comment documenting that assumption.

File: aap_gateway_api/tests/views/api/v1/ping/test_ping.py [hunk 3]
```diff
@@ -10,8 +10,9 @@
 
 
 @pytest.mark.django_db
+@mock.patch("aap_gateway_api.views.api.v1.ping.PingView._check_dispatcherd")
```
Comment (Dostonbek1) [inline aap_gateway_api/tests/views/api/v1/ping/test_ping.py:13]: This `_check_dispatcherd` mock is correctly added here, but the same mock is missing from `test_ping_db_down`, `test_ping_proxy_exception`, and `test_ping_proxy_non_200`. Since the dispatcherd check now runs unconditionally (outside the main DB/proxy try/except), those tests will attempt a real dispatcherd connection. This creates environment-dependent behavior and a test isolation gap.

Consider adding this decorator to those tests as well, or using a module-level `autouse` fixture to mock dispatcherd by default.

**Review by Dostonbek1 — COMMENTED** [review]
## 📋 Code Review Summary

This PR adds a `dispatcherd_connected` field to the ping endpoint, providing visibility into dispatcherd health. The approach is clean — the check is informational only and correctly isolated from the overall status determination. A few items worth addressing around logging and test isolation.

**Files Reviewed**: 3 files
**Comments Posted**: 3 review comments

### 🔍 Issues Found
- 0 security/safety issues
- 1 correctness/logic issue (missing mock on existing tests)
- 2 code quality suggestions (logging, return value handling)

### Overall Assessment
Looks good overall — small, focused change with proper test coverage for the new feature. The main concern is that existing tests now have an unmocked dependency on `_check_dispatcherd`, which could cause environment-dependent behavior. Adding logging and mocking those tests would make this merge-ready.

### General Feedback
- Good decision to make dispatcherd status informational-only, not affecting the overall status
- The deferred import pattern in `_check_dispatcherd` is appropriate
- Test coverage for the new feature is solid with both up/down scenarios

File: aap_gateway_api/views/api/v1/ping.py [hunk 4]
```diff
@@ -60,5 +65,12 @@ def get(self, request):
                 {'status': STATUS_DEGRADED, 'db_connected': False, 'db_exception': type(e).__name__, 'proxy_exception_type': 'Skipped, DB unavailable.'}
             )
 
+        # Dispatcherd check is informational only — does not affect overall status
+        try:
+            self._check_dispatcherd()
+            response['dispatcherd_connected'] = True
+        except Exception:
```
Comment (bhavenst) [inline aap_gateway_api/views/api/v1/ping.py:76]: Probably good to log the response in the negative case..

File: aap_gateway_api/views/api/v1/ping.py [hunk 5]
```diff
@@ -60,5 +65,12 @@ def get(self, request):
                 {'status': STATUS_DEGRADED, 'db_connected': False, 'db_exception': type(e).__name__, 'proxy_exception_type': 'Skipped, DB unavailable.'}
             )
 
+        # Dispatcherd check is informational only — does not affect overall status
+        try:
+            self._check_dispatcherd()
```
Comment (bhavenst) [inline aap_gateway_api/views/api/v1/ping.py:74]: I agree with coderabbit that it's probably good to verify the actual response here.

File: aap_gateway_api/tests/views/api/v1/ping/test_ping.py (hunk 3 above)
Comment (tznamena) [reply aap_gateway_api/tests/views/api/v1/ping/test_ping.py:13]: Added.

File: aap_gateway_api/views/api/v1/ping.py (hunk 2 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:30]: Updated the docstring to mention that the alive command either gets a reply or raises an exception.

File: aap_gateway_api/views/api/v1/ping.py (hunk 4 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:76]: Added.

File: aap_gateway_api/views/api/v1/ping.py (hunk 5 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:74]: The contract is that either we get a response or it throws an exception.

File: aap_gateway_api/views/api/v1/ping.py (hunk 1 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:78]: Added.

File: aap_gateway_api/views/api/v1/ping.py [hunk 6]
```diff
@@ -11,6 +12,8 @@
 from aap_gateway_api.version import get_aap_version
 from aap_gateway_api.views.api.v1.common import AnsibleBaseView
 
+logger = logging.getLogger(__name__)
```
Comment (john-westcott-iv) [inline aap_gateway_api/views/api/v1/ping.py]: **minor:** The gateway convention is `aap.gateway.*` for logger names, not `__name__`. Using `__name__` here produces `aap_gateway_api.views.api.v1.ping` (underscores) which is inconsistent with every other view in the codebase.

```suggestion
logger = logging.getLogger('aap.gateway.views.api.v1.ping')
```

File: aap_gateway_api/views/api/v1/ping.py [hunk 7]
```diff
@@ -20,6 +23,12 @@ def _check_db(self):
         with connection.cursor() as cursor:
             cursor.execute("SELECT 1")
 
+    def _check_dispatcherd(self):
+        """Raises on failure — the alive command either gets a reply or throws."""
+        from dispatcherd.factories import get_control_from_settings
+
+        get_control_from_settings().control_with_reply("alive", timeout=getattr(settings, "DISPATCHERD_HEALTH_CHECK_TIMEOUT", 5))
+
```
Comment (john-westcott-iv) [inline aap_gateway_api/views/api/v1/ping.py:31]: **discussion:** This is consistent with how `PING_PAGE_CHECK_TIMEOUT` is handled in this same file (`getattr(settings, ...)`). However, the nearby `status.py` view uses `get_preference_value` from the dynamic preferences system for runtime-configurable values. Worth considering whether this timeout should be a dynamic preference (changeable without restart) rather than a static setting — especially if operators need to tune it in response to dispatcherd latency without a gateway restart.

File: aap_gateway_api/tests/views/api/v1/ping/test_ping.py [hunk 8]
```diff
@@ -75,3 +80,31 @@ def test_ping_proxy_non_200(request, unauthenticated_api_client, service_cluster
     assert response.status_code == 200
     assert response.data['status'] == STATUS_DEGRADED
     assert response.data['proxy_status_code'] == 500
+
+
+@pytest.mark.django_db
+@mock.patch("aap_gateway_api.views.api.v1.ping.PingView._check_dispatcherd")
+@mock.patch("aap_gateway_api.views.api.v1.ping.requests.request")
+def test_ping_dispatcherd_up(request, mock_dispatcherd, unauthenticated_api_client):
```
Comment (john-westcott-iv) [inline aap_gateway_api/tests/views/api/v1/ping/test_ping.py]: **nit:** This test is nearly identical to `test_ping_all_up` above, which already asserts `dispatcherd_connected is True` (line 25). The only difference is that `test_ping_all_up` also checks `version` and other fields. Could be removed in favor of the assertion already in `test_ping_all_up`, or differentiated to test something unique (e.g., DB down + dispatcherd up). Author's discretion.

**Review by john-westcott-iv — COMMENTED** [review]
## Code Review: AAP-65395 — Add dispatcherd health check to ping endpoint

**Verdict**: READY_FOR_HUMAN_REVIEW
**Score**: 10/10 (Functionality 10, Security 10, Quality 10)

Clean, well-scoped change. The dispatcherd health check is correctly isolated as informational-only, properly guarded with its own try/except, and does not affect overall ping status. Security is clean — hardcoded `alive` command, boolean-only response, no user input flows into the control call.

**Inline comments:** 1 minor (logger naming convention), 1 discussion (dynamic preferences vs static settings for timeout), 1 nit (test redundancy).

**Additional observation:** The `logger.warning` with `exc_info=True` on line 77 may be noisy in production. The ping endpoint is polled by load balancers every few seconds — if dispatcherd is down, this generates a WARNING with full traceback on every poll (potentially hundreds per hour). The existing DB and proxy checks do not log on failure. Consider `logger.debug` or `logger.info` to stay consistent and avoid log flooding.

File: aap_gateway_api/views/api/v1/ping.py (hunk 7 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:31]: Keeping `getattr(settings, ...)` for now. Rationale:

- **Consistency within the file** — `PING_PAGE_CHECK_TIMEOUT` on line 33 uses the same `getattr` pattern.
- **No DB query per ping** — `get_preference_value` would add a DB read on every `/ping` call. The ping endpoint should be lightweight since it's called frequently by health checks and load balancers.
- **Rarely needs tuning** — the 5s default is reasonable for most environments and unlikely to need runtime changes.

If this becomes a real operational need (e.g., operators frequently tuning dispatcherd timeout due to latency), it can be migrated to a dynamic preference later.

File: aap_gateway_api/views/api/v1/ping.py (hunk 6 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py]: Fixed.

File: aap_gateway_api/views/api/v1/ping.py (hunk 7 above)
Comment (tznamena) [reply aap_gateway_api/views/api/v1/ping.py:31]: @john-westcott-iv Do you agree with the rationale above?

File: aap_gateway_api/tests/views/api/v1/ping/test_ping.py (hunk 8 above)
Comment (tznamena) [reply aap_gateway_api/tests/views/api/v1/ping/test_ping.py]: Consolidated the two tests.

File: aap_gateway_api/views/api/v1/ping.py (hunk 7 above)
Comment (john-westcott-iv) [reply aap_gateway_api/views/api/v1/ping.py:31]: Yea, I'm good with that. Go ahead and leave it as a setting.

**Review by john-westcott-iv — APPROVED** [review]
All prior review comments addressed. Logger convention fixed (`aap.gateway.*`), `_check_dispatcherd` mocked in all existing tests, docstring documents the alive exception-based contract, warning logged on failure, duplicate test consolidated. LGTM.

**Review by Dostonbek1 — APPROVED** [review]
All review feedback has been addressed:

- ✅ Logging added for dispatcherd health check failures (`logger.warning` with `exc_info=True`)
- ✅ Docstring added documenting the exception-based contract (`"Raises on failure — the alive command either gets a reply or throws."`)
- ✅ `_check_dispatcherd` mock added to all existing tests for proper isolation

LGTM — clean, well-tested change. Nice work!

**Review by bhavenst — APPROVED** [review]
